spinoso-string/src/enc/ascii/impls.rs

```diff
@@ -0,0 +1,180 @@
+use alloc::borrow::Cow;
+use alloc::string::String;
+use alloc::vec::Vec;
+use core::borrow::{Borrow, BorrowMut};
+use core::fmt::Arguments;
+use core::ops::{Deref, DerefMut};
+#[cfg(feature = "std")]
+use std::io::{IoSlice, Result, Write};
+
+use super::AsciiString;
+
+#[cfg(feature = "std")]
+impl Write for AsciiString {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        self.inner.write(buf)
+    }
+
+    #[inline]
+    fn write_all(&mut self, buf: &[u8]) -> Result<()> {
+        self.inner.write_all(buf)
+    }
+
+    #[inline]
+    fn write_fmt(&mut self, fmt: Arguments<'_>) -> Result<()> {
+        self.inner.write_fmt(fmt)
+    }
+
+    #[inline]
+    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> Result<usize> {
+        self.inner.write_vectored(bufs)
+    }
+
+    #[inline]
+    fn flush(&mut self) -> Result<()> {
+        self.inner.flush()
+    }
+}
+
+impl Extend<u8> for AsciiString {
+    #[inline]
+    fn extend<I: IntoIterator<Item = u8>>(&mut self, iter: I) {
+        self.inner.extend(iter.into_iter());
+    }
+}
+
+impl<'a> Extend<&'a u8> for AsciiString {
+    #[inline]
+    fn extend<I: IntoIterator<Item = &'a u8>>(&mut self, iter: I) {
+        self.inner.extend(iter.into_iter().copied());
+    }
+}
+
+impl<'a> Extend<&'a mut u8> for AsciiString {
+    #[inline]
+    fn extend<I: IntoIterator<Item = &'a mut u8>>(&mut self, iter: I) {
+        self.inner.extend(iter.into_iter().map(|byte| *byte));
+    }
+}
+
+impl From<Vec<u8>> for AsciiString {
+    #[inline]
+    fn from(content: Vec<u8>) -> Self {
+        Self::new(content)
+    }
+}
+
+impl<'a> From<&'a [u8]> for AsciiString {
+    #[inline]
+    fn from(content: &'a [u8]) -> Self {
+        Self::new(content.to_vec())
+    }
+}
+
+impl<'a> From<&'a mut [u8]> for AsciiString {
+    #[inline]
+    fn from(content: &'a mut [u8]) -> Self {
+        Self::new(content.to_vec())
+    }
+}
+
+impl<'a> From<Cow<'a, [u8]>> for AsciiString {
+    #[inline]
+    fn from(content: Cow<'a, [u8]>) -> Self {
+        Self::new(content.into_owned())
+    }
+}
+
+impl From<String> for AsciiString {
+    #[inline]
+    fn from(s: String) -> Self {
+        Self::new(s.into_bytes())
+    }
+}
+
+impl From<&str> for AsciiString {
+    #[inline]
+    fn from(s: &str) -> Self {
+        Self::new(s.as_bytes().to_vec())
+    }
+}
+
+impl From<AsciiString> for Vec<u8> {
+    #[inline]
+    fn from(s: AsciiString) -> Self {
+        s.to_vec()
```

Comment:
should this be `s.into_vec()`? It looks like `self.to_vec()` is coming from the deref to `&[u8]`. This means we're allocating when we don't need to be since the `From` impl consumes `s`.

```suggestion
        s.into_vec()
```

Reply:
Ooo, good catch. Now I'm super curious how rust even knows this. e.g. `to_vec` could be implemented on any type, so does it just check the `Deref::Target` to see if it implements something that has a `to_vec` function, and then uses that if all else fails? Or am I missing something here?